File: ios/ReactNativeArViewModule.swift
import ExpoModulesCore

public class ReactNativeArViewModule: Module {
    public func definition() -> ModuleDefinition {
        Name("ReactNativeArView")

        Events(
            "onTrackingStateChange",
            "onPlaneDetected",
            "onModelLoaded",
            "onModelPlaced",
            "onModelError",
            "onSceneChange",
            "onARError"
        )

        AsyncFunction("pushScene") { (models: [[String: Any]]) in
            let configs = models.map { ModelConfig.from($0) }
            await withCheckedContinuation { continuation in
                DispatchQueue.main.async {
                    Self.getView()?.pushScene(configs)
                    continuation.resume()
                }
            }
        }

        AsyncFunction("popScene") { () -> Bool in
            return await withCheckedContinuation { continuation in
                DispatchQueue.main.async {
                    let result = Self.getView()?.popScene() ?? false
                    continuation.resume(returning: result)
                }
            }
        }

        AsyncFunction("replaceScene") { (models: [[String: Any]]) in
            let configs = models.map { ModelConfig.from($0) }
            await withCheckedContinuation { continuation in
                DispatchQueue.main.async {
                    Self.getView()?.replaceScene(configs)
                    continuation.resume()
                }
            }
        }

        AsyncFunction("popToTop") { () in
            await withCheckedContinuation { continuation in
                DispatchQueue.main.async {
                    Self.getView()?.popToTop()
                    continuation.resume()
                }
            }
        }

        View(ReactNativeArView.self) {
            Events(
                "onTrackingStateChange",
                "onPlaneDetected",
                "onModelLoaded",
                "onModelPlaced",
                "onModelError",
                "onSceneChange",
                "onARError"
            )
        }
    }

    private static func getView() -> ReactNativeArView? {
        return ReactNativeArView.currentInstance
    }
}

File: android/src/main/java/expo/modules/arview/ReactNativeArViewModule.kt
package expo.modules.arview

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise
import android.os.Handler
import android.os.Looper

class ReactNativeArViewModule : Module() {
    private val uiHandler = Handler(Looper.getMainLooper())

    private fun runOnUi(promise: Promise, block: () -> Any?) {
        uiHandler.post {
            try {
                promise.resolve(block())
            } catch (e: Exception) {
                promise.reject("ERR_UI_THREAD", e.message, e)
            }
        }
    }

    override fun definition() = ModuleDefinition {
        Name("ReactNativeArView")

        Events(
            "onTrackingStateChange",
            "onPlaneDetected",
            "onModelLoaded",
            "onModelPlaced",
            "onModelError",
            "onSceneChange",
            "onARError"
        )

        AsyncFunction("pushScene") { models: List<Map<String, Any?>>, promise: Promise ->
            val configs = models.map { ModelConfig.fromMap(it) }
            runOnUi(promise) { getView()?.pushScene(configs) }
        }

        AsyncFunction("popScene") { promise: Promise ->
            runOnUi(promise) { getView()?.popScene() ?: false }
        }

        AsyncFunction("replaceScene") { models: List<Map<String, Any?>>, promise: Promise ->
            val configs = models.map { ModelConfig.fromMap(it) }
            runOnUi(promise) { getView()?.replaceScene(configs) }
        }

        AsyncFunction("popToTop") { promise: Promise ->
            runOnUi(promise) { getView()?.popToTop() }
        }

        View(ReactNativeArView::class) {
            Events(
                "onTrackingStateChange",
                "onPlaneDetected",
                "onModelLoaded",
                "onModelPlaced",
                "onModelError",
                "onSceneChange",
                "onARError"
            )

            Prop("planeDetection") { view: ReactNativeArView, value: String? ->
                view.setPlaneDetection(value ?: "horizontal_and_vertical")
            }

            // Gesture listener is set up once in ReactNativeArView init block
        }
    }

    private fun getView(): ReactNativeArView? {
        return ReactNativeArView.currentInstance?.get()
    }
}
